packages/astro/src/core/sync/index.ts

```diff
@@ -28,73 +22,92 @@ import {
 import type { Logger } from '../logger/core.js';
 import { formatErrorMessage } from '../messages.js';
 import { ensureProcessNodeEnv } from '../util.js';
-
-export type ProcessExit = 0 | 1;
+import { createNodeLogger } from '../config/logging.js';
+import { resolveConfig } from '../config/config.js';
+import { createSettings } from '../config/settings.js';
+import { telemetry } from '../../events/index.js';
+import { eventCliSession } from '../../events/session.js';
+import { runHookConfigSetup } from '../../integrations/hooks.js';
 
 export type SyncOptions = {
 	/**
 	 * @internal only used for testing
 	 */
 	fs?: typeof fsMod;
-};
-
-export type SyncInternalOptions = SyncOptions & {
 	logger: Logger;
+	settings: AstroSettings;
+	skip?: {
+		// Must be skipped in dev
+		content?: boolean;
+	};
 };
 
 type DBPackage = {
 	typegen?: (args: Pick<AstroConfig, 'root' | 'integrations'>) => Promise<void>;
 };
 
+export async function syncInlineConfig({
+	inlineConfig,
+	fs,
+	telemetry: _telemetry = false,
+}: { inlineConfig: AstroInlineConfig; fs?: typeof fsMod; telemetry?: boolean }) {
+	const logger = createNodeLogger(inlineConfig);
+	const { astroConfig, userConfig } = await resolveConfig(inlineConfig ?? {}, 'sync');
+	if (_telemetry) {
+		telemetry.record(eventCliSession('sync', userConfig));
+	}
+	let settings = await createSettings(astroConfig, inlineConfig.root);
+	settings = await runHookConfigSetup({
+		command: 'build',
+		settings,
+		logger,
+	});
+	return await sync({ settings, logger, fs });
+}
+
 /**
  * Generates TypeScript types for all Astro modules. This sets up a `src/env.d.ts` file for type inferencing,
  * and defines the `astro:content` module for the Content Collections API.
  *
  * @experimental The JavaScript API is experimental
  */
-export default async function sync(
-	inlineConfig: AstroInlineConfig,
-	options?: SyncOptions
-): Promise<ProcessExit> {
+export default async function sync({
+	logger,
+	fs = fsMod,
+	settings,
+	skip,
+}: SyncOptions): Promise<void> {
 	ensureProcessNodeEnv('production');
```

Comment:
This should be called at the very first before we do any work (as soon as we identify `sync()` or `astro sync` is being called. So this should be moved into the above function.